## Prediksi hasil match berdasarkan hero yang di pilih di Dota 2

Dota adalah sebuah game ber genre MOBA dengan platform PC yang sangat populer.  
Di sebuah game Dota, terdapat 2 tim yang berlawanan, yaitu Radiant dan Dire. Masing masing team mempunyai 5 buah player.
Dota mempunyai banyak hal yang dapat mempengaruhi hasil match, seperti skill, strategi, item, timing dan sebagainya.

Namun di tugas ini, kita hanya akan membahas hero yang di pilih oleh satu sisi.
Dota 2 mempunyai total hero sebanyak 123 hero.
Satu tim akan memilih 5 hero untuk dimainkan, dan kita akan menghitung presentase kemenangan tim tersebut berdasarkan hero yang di pilih.

Kita mempunyai data 10.000 match dari Dota 2 yang berlangsung di seluruh penjuru dunia.
Match tersebut tersebar dari beberapa patch yang berbeda, dan juga berbeda tahun.
Di setiap patch, ada hero baru, item baru, dan juga nerf hero atau item.
Perbedaan antar patch ini akan mempengaruhi prediksi algoritma kita.

Data ini diambil langsung dari Dota 2 API.

In [12]:
import pandas as pd

df = pd.read_csv('global.csv')
df.head()

Match ID  League         Start Date  Duration (s) Duration (mm:ss)  \
0  4230890141   10288  21 Nov 2018 08:00          2022            33:42   
1  4230971959   10288  21 Nov 2018 09:04          1576            26:16   
2  4232651707   10288  22 Nov 2018 07:59          2158            35:58   
3  4232732529   10288  22 Nov 2018 09:04          1700            28:20   
4  4234538318   10288  23 Nov 2018 10:08          1371            22:51   

  Radiant Team   Team A                                      Team A Heroes  \
0       Team A  PSG.LGD  Juggernaut  Silencer  Rubick  Magnus  Monkey King   
1       Team A  PSG.LGD   Dragon Knight  Doom  Undying  Rubick  Arc Warden   
2       Team A  PSG.LGD  Phantom Lancer  Shadow Shaman  Doom  Outworld ...   
3       Team A  PSG.LGD     Lich  Bounty Hunter  Lone Druid  Rubick  Slark   
4       Team A  PSG.LGD       Sand King  Lich  Gyrocopter  Io  Monkey King   

            Team B                                      Team B Heroes  Winner  
0       Team Aster  Luna  Dragon Knight  Ursa  Shadow Demon  Trean...  Team A  
1       Team Aster  Shadow Shaman  Faceless Void  Alchemist  Brewm...  Team A  
2  INVICTUS GAMING             Bane  Leshrac  Omniknight  Lycan  Tusk  Team A  
3  INVICTUS GAMING  Juggernaut  Kunkka  Treant Protector  Magnus  ...  Team A  
4    Taichi Gaming   Bane  Shadow Shaman  Queen of Pain  Ursa  Visage  Team A

### Step 1 : Pilih kolom yang dibutuhkan

In [13]:
df1 = pd.DataFrame(df, columns=["Match ID", "Team A Heroes", "Winner"])
df1.head()

Match ID                                      Team A Heroes  Winner
0  4230890141  Juggernaut  Silencer  Rubick  Magnus  Monkey King  Team A
1  4230971959   Dragon Knight  Doom  Undying  Rubick  Arc Warden  Team A
2  4232651707  Phantom Lancer  Shadow Shaman  Doom  Outworld ...  Team A
3  4232732529     Lich  Bounty Hunter  Lone Druid  Rubick  Slark  Team A
4  4234538318       Sand King  Lich  Gyrocopter  Io  Monkey King  Team A

### Step 2 : Rename kolom

In [14]:
#rename columns

df2 = pd.DataFrame(df1)
df2.rename(columns={'Match ID': 'match_id', 'Team A Heroes': 'hero1', 'Winner': 'result'}, inplace=True)
df2.head()

match_id                                              hero1  result
0  4230890141  Juggernaut  Silencer  Rubick  Magnus  Monkey King  Team A
1  4230971959   Dragon Knight  Doom  Undying  Rubick  Arc Warden  Team A
2  4232651707  Phantom Lancer  Shadow Shaman  Doom  Outworld ...  Team A
3  4232732529     Lich  Bounty Hunter  Lone Druid  Rubick  Slark  Team A
4  4234538318       Sand King  Lich  Gyrocopter  Io  Monkey King  Team A

### Step 3 : Buat kolom baru untuk 5 hero di satu tim

In [15]:
df3 = pd.DataFrame(df2)
df3.insert(2, 'hero2', 0)
df3.insert(3, 'hero3', 0)
df3.insert(4, 'hero4', 0)
df3.insert(5, 'hero5', 0)

df3.head()

match_id                                              hero1  hero2  \
0  4230890141  Juggernaut  Silencer  Rubick  Magnus  Monkey King      0   
1  4230971959   Dragon Knight  Doom  Undying  Rubick  Arc Warden      0   
2  4232651707  Phantom Lancer  Shadow Shaman  Doom  Outworld ...      0   
3  4232732529     Lich  Bounty Hunter  Lone Druid  Rubick  Slark      0   
4  4234538318       Sand King  Lich  Gyrocopter  Io  Monkey King      0   

   hero3  hero4  hero5  result  
0      0      0      0  Team A  
1      0      0      0  Team A  
2      0      0      0  Team A  
3      0      0      0  Team A  
4      0      0      0  Team A

### Step 4 : Proses data lebih lanjut

- Rename data di kolom result menjadi 1 dan 0. 1 berarti menang, 0 berarti kalah
- Split data di hero1, lalu sort berdasarkan alphabet
- hapus space yang ada di depan dan belakang nama hero
- lalu masukkan ke dalam kolom hero 2-5

In [16]:
#loop for each row, adding heroes to new columns

df4 = pd.DataFrame(df3)

#change result column to 1 and 0. 1 means winning, 0 means losing
for i in range(len(df4)): 
    if (df4.loc[i,'result'] == "Team A"):
        df4.loc[i,'result'] = 1
    elif (df4.loc[i,'result'] == "Team B"):
        df4.loc[i,'result'] = 0
        
    data=df4.loc[i,'hero1']
    data=data.split("  ")
    data.sort()
    
    df4.loc[i,'hero1']=data[0].strip()
    df4.loc[i,'hero2']=data[1].strip()
    df4.loc[i,'hero3']=data[2].strip()
    df4.loc[i,'hero4']=data[3].strip()
    df4.loc[i,'hero5']=data[4].strip()
    
df4.head()

match_id          hero1         hero2              hero3           hero4  \
0  4230890141     Juggernaut        Magnus        Monkey King          Rubick   
1  4230971959     Arc Warden          Doom      Dragon Knight          Rubick   
2  4232651707           Doom  Earth Spirit  Outworld Devourer  Phantom Lancer   
3  4232732529  Bounty Hunter          Lich         Lone Druid          Rubick   
4  4234538318     Gyrocopter            Io               Lich     Monkey King   

           hero5 result  
0       Silencer      1  
1        Undying      1  
2  Shadow Shaman      1  
3          Slark      1  
4      Sand King      1

### Step 5 : Buat ID untuk hero yang unik

- Kita akan mengambil semua hero yang ada di kolom hero1-5, dari semua match.
- Lalu kita hapus duplikat yang ada di dalam data.
- Selanjutnya kita masukkan semua data string ke dalam sebuah list, dan urutkan berdasarkan alphabet

hasilnya hero yang mempunyai awalan A akan berada di index/id bagian awal
Sebagai contoh :  
Abaddon mempunyai id 0
Alchemist mempunyai id 1  
Bounty Hunter mempunyai id 10


In [17]:
#make unique ID for each hero
df5 = pd.DataFrame(df4, columns=["hero1", "hero2", "hero3", "hero4", "hero5"])

allheroes=[]
for i in range(len(df5)):
    allheroes.append(df5.loc[i,'hero1'])
    allheroes.append(df5.loc[i,'hero2'])
    allheroes.append(df5.loc[i,'hero3'])
    allheroes.append(df5.loc[i,'hero4'])
    allheroes.append(df5.loc[i,'hero5'])
    
allheroes = list(dict.fromkeys(allheroes))
allheroes = sorted(allheroes)
print(allheroes)
print(len(allheroes))

['Abaddon', 'Alchemist', 'Ancient Apparition', 'Anti-Mage', 'Arc Warden', 'Axe', 'Bane', 'Batrider', 'Beastmaster', 'Bloodseeker', 'Bounty Hunter', 'Brewmaster', 'Bristleback', 'Broodmother', 'Centaur Warrunner', 'Chaos Knight', 'Chen', 'Clinkz', 'Clockwerk', 'Crystal Maiden', 'Dark Seer', 'Dark Willow', 'Dawnbreaker', 'Dazzle', 'Death Prophet', 'Disruptor', 'Doom', 'Dragon Knight', 'Drow Ranger', 'Earth Spirit', 'Earthshaker', 'Elder Titan', 'Ember Spirit', 'Enchantress', 'Enigma', 'Faceless Void', 'Grimstroke', 'Gyrocopter', 'Hoodwink', 'Huskar', 'Invoker', 'Io', 'Jakiro', 'Juggernaut', 'Keeper of the Light', 'Kunkka', 'Legion Commander', 'Leshrac', 'Lich', 'Lifestealer', 'Lina', 'Lion', 'Lone Druid', 'Luna', 'Lycan', 'Magnus', 'Marci', 'Mars', 'Medusa', 'Meepo', 'Mirana', 'Monkey King', 'Morphling', 'Naga Siren', "Nature's Prophet", 'Necrophos', 'Night Stalker', 'Nyx Assassin', 'Ogre Magi', 'Omniknight', 'Oracle', 'Outworld Devourer', 'Pangolier', 'Phantom Assassin', 'Phantom Lancer

### Step 6 : Ubah data nama hero menjadi angka (id)

- Loop semua data di kolom hero1-5
- jika nama hero ada di list allheroes, ambil index/id dari hero tersebut
- gantikan nama hero dengan index/id hero tersebut

In [18]:
#replace hero names with hero id

df6 = pd.DataFrame(df4)

for i in range(len(df6)):
    x=(df6.loc[i,'hero1'])
    xx=allheroes.index(x, 0, len(allheroes))
    df6.loc[i,'hero1'] = xx
    
    y=(df6.loc[i,'hero2'])
    yy=allheroes.index(y, 0, len(allheroes))
    df6.loc[i,'hero2'] = yy
    
    z=(df6.loc[i,'hero3'])
    zz=allheroes.index(z, 0, len(allheroes))
    df6.loc[i,'hero3'] = zz
    
    c=(df6.loc[i,'hero4'])
    cc=allheroes.index(c, 0, len(allheroes))
    df6.loc[i,'hero4'] = cc
    
    d=(df6.loc[i,'hero5'])
    dd=allheroes.index(d, 0, len(allheroes))
    df6.loc[i,'hero5'] = dd

df6.head()

match_id hero1 hero2 hero3 hero4 hero5 result
0  4230890141    43    55    61    83    88      1
1  4230971959     4    26    27    83   109      1
2  4232651707    26    29    71    74    87      1
3  4232732529    10    48    52    83    91      1
4  4234538318    37    41    48    61    84      1

In [19]:
df6.to_csv(r'./finaldata.csv', index=False)

### Step 7 : load ke pandas, hilangkan kolom matchid

In [20]:
import pandas as pd

data = pd.read_csv('./finaldata.csv')
data = data.drop(['match_id'], axis=1)
data.to_csv(r'./finaldata.csv', index=False)

data.head()

hero1  hero2  hero3  hero4  hero5  result
0     43     55     61     83     88       1
1      4     26     27     83    109       1
2     26     29     71     74     87       1
3     10     48     52     83     91       1
4     37     41     48     61     84       1

### Step 8 : Split data menjadi data train dan data test

In [21]:
from sklearn.model_selection import train_test_split

X = data.drop('result',axis=1)
y = data["result"]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state = 1)

### Step 9 : buat model berdasarkan algoritma naive bayes

**Naive Bayes Gaussian**  
 Algoritma ini mengasumsikan bahwa data mengikuti distribusi normal dan menghitung kemungkinan setiap kelas berdasarkan rata-rata dan standar deviasi fitur untuk kelas tersebut.

**Naive Bayes Multinomial**  
 Algoritma ini digunakan untuk tugas klasifikasi di mana fitur-fiturnya diskrit dan multinomial (misalnya, hitungan kata dalam dokumen). Algoritma ini menghitung kemungkinan setiap kelas berdasarkan frekuensi setiap fitur dalam data latihan.

**Naive Bayes Bernoulli**  
 Algoritma ini mirip dengan algoritma naive Bayes multinomial, tetapi digunakan untuk tugas klasifikasi di mana fitur-fiturnya biner (yaitu, hanya memiliki dua nilai). Algoritma ini menghitung kemungkinan setiap kelas berdasarkan adanya atau tidaknya setiap fitur dalam data latihan.

 **Naive Bayes Complementary**  
 adalah algoritma yang didasarkan pada prinsip yang sama seperti naive Bayes, tetapi berbeda dalam beberapa cara khusus. Sebagai contoh, mungkin ada varian dari algoritma naive Bayes yang menggunakan distribusi yang berbeda untuk memodelkan data (misalnya, distribusi gamma bukan distribusi normal) atau yang melepaskan asumsi independensi antar fitur.

In [22]:
from sklearn.naive_bayes import GaussianNB, BernoulliNB, MultinomialNB, ComplementNB
from sklearn.metrics import accuracy_score
import pickle5 as pickle

nb = GaussianNB()
nb.fit(X_train, y_train)
nb_predict = nb.predict(X_test)

nb_b = BernoulliNB()
nb_b.fit(X_train, y_train)
nb_b_predict = nb_b.predict(X_test)

nb_m = MultinomialNB()
nb_m.fit(X_train, y_train)
nb_m_predict = nb_m.predict(X_test)

nb_c = ComplementNB()
nb_c.fit(X_train, y_train)
nb_c_predict = nb_c.predict(X_test)

nb_acc_score = accuracy_score(y_test, nb_predict)
nb_b_score = accuracy_score(y_test, nb_b_predict)
nb_m_score = accuracy_score(y_test, nb_m_predict)
nb_c_score = accuracy_score(y_test, nb_c_predict)

with open('nb_gaussian.pkl', 'wb') as file:
  pickle.dump(nb, file)
  
with open('nb_bernoulli.pkl', 'wb') as file:
  pickle.dump(nb_b, file)
  
with open('nb_multinomial.pkl', 'wb') as file:
  pickle.dump(nb_m, file)
  
with open('nb_complementary.pkl', 'wb') as file:
  pickle.dump(nb_c, file)

print(f'naive bayes : {nb_acc_score}')
print(f'bernoulli : {nb_b_score}')
print(f'multinomial : {nb_m_score}')
print(f'complementary : {nb_c_score}')

naive bayes : 0.5988142292490118
bernoulli : 0.5988142292490118
multinomial : 0.5691699604743083
complementary : 0.5079051383399209


### Step 10 : Buat model berdasarkan algoritma SVM

**Apa itu SVM?**

Support vector machines (SVMs) adalah sebuah algoritma yang digunakan untuk klasifikasi dan regresi. SVMs adalah salah satu algoritma yang paling populer untuk machine learning, karena algoritma ini sangat fleksibel dan dapat digunakan untuk klasifikasi linier dan non-linier, regresi, dan deteksi outlier.

Parameter yang digunakan adalah :

- Kernel, adalah parameter yang paling penting dari SVM. Kernel adalah fungsi yang digunakan untuk menghitung jarak antara dua data. Kernel yang digunakan adalah linear, polynomial, dan RBF.

- regularization (C), adalah parameter yang digunakan untuk mengatur trade-off antara memperbolehkan model untuk menyesuaikan data latih dengan baik dan mencegah overfitting. Semakin besar nilai C, maka model akan mencoba untuk menyesuaikan data latih dengan baik, yang dapat menyebabkan overfitting. Semakin kecil nilai C, maka model akan lebih sensitif terhadap noise di data latih, yang dapat meningkatkan generalisasi ke data baru.

- gamma/degree, adalah parameter spesifik untuk model polynomial dan RBF yang digunakan untuk mengontrol lebar dari filter gauss. Semakin besar nilai gamma, maka model akan lebih sensitif terhadap perubahan data latih, yang dapat menyebabkan overfitting. Semakin kecil nilai gamma, maka model akan lebih sensitif terhadap noise di data latih, yang dapat meningkatkan generalisasi ke data baru.

In [23]:
# import necessary libraries
from sklearn import datasets
from sklearn import svm
from sklearn.model_selection import train_test_split

# split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=0)

# create an SVM model and train it on the training data
model = svm.SVC(kernel="linear", C=5)
model.fit(X_train, y_train)

# evaluate the model on the test data
accuracy = model.score(X_test, y_test)
print("Accuracy: {:.2f}".format(accuracy))

with open('svm_linear_c5.pkl', 'wb') as file:
  pickle.dump(model, file)

Accuracy: 0.59


### Kesimpulan :

Setelah kita menganalisa 10.000 professional dota2 match menggunakan dua algoritma yang berbeda, ternyata hasilnya sama sama menunjukan 59% akurasi.  

kesimpulan utamanya adalah Dota adalah game yang sangat kompleks. Kemenangan suatu game ditentukan oleh banyak faktor seperti  
    - skill player  
    - strategi tim  
    - item yang di pilih  
    - power spike timing  
    - **bejo**  

Hero yang dipilih hanyalah satu dari sekian banyak faktor yang mempengaruhi kemenangan sebuah tim, dan tidak akan mencerminkan kemenangan atau kekalahan tim secara sempurna.
